Porting pyvmi to CFFI

Wenzel opened the pull request:
Hi,

this is a big WIP to create a Python API for `LibVMI` using `CFFI`.

The API is not complete, i only implemented enough functions to get a the process listing test working, which i also reimplemented in `test_libvmi.py`

Drop some comments.

To use it:
1. `python3 build_libvmi.py`
2. `./test_libvmi.py <vm_name>`

libvmi-jenkins commented:
Can one of the admins verify this patch?

Wenzel commented:
Test output:

~~~
 wenzel@localhost  ~/Projets/libvmi/tools/pyvmi   pyvmi_cffi ✘ ✹ ✭  ./test_libvmi.py nitro_win7x64                                                                                         22:25:22  ✘ 130 
xencall: error: Could not obtain handle on privileged command interface: No such file or directory
VMI_ERROR: Failed to open libxc interface.
LibVMI Version 0.11.0
LibVMI Driver Mode 1
--completed driver init.
--got id from name (nitro_win7x64 --> 6)
**set image_type = nitro_win7x64
--libvirt version 3002001
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "pmemaccess", "arguments": {"path": "/tmp/vmi0dFSaw"}}'
--kvm: using custom patch for fast memory access
**set allocated_ram_size = 59700000, max_physical_address = 0x59700000
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**set pae = 1
**set pse = 1
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**set lme = 1
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
**IA-32e paging
**sanity checking cr3 = 0x0000000000187000
--looking for config file at /home/wenzel/ownCloud/Projets/libvmi/tools/pyvmi/libvmi.conf
--looking for config file at /home/wenzel/etc/libvmi.conf
--looking for config file at /etc/libvmi.conf
**Using config file at /etc/libvmi.conf
**set os_type to Windows.
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
LibVMI Suggestion: set win_ntoskrnl=0x261f000 in libvmi.conf for faster startup.
LibVMI Suggestion: set win_kdbg=0x1e9070 in libvmi.conf for faster startup.
LibVMI Suggestion: set win_kdvb=0xfffff80002808070 in libvmi.conf for faster startup.
Process listing for VM nitro_win7x64 (id: 6)
[4] System (struct addr:0xfffffa8001167040)
[232] smss.exe (struct addr:0xfffffa800198c950)
[304] csrss.exe (struct addr:0xfffffa80021465f0)
[352] wininit.exe (struct addr:0xfffffa8002169700)
[360] csrss.exe (struct addr:0xfffffa800216c060)
[388] winlogon.exe (struct addr:0xfffffa8002175830)
[448] services.exe (struct addr:0xfffffa80020eeb30)
[456] lsass.exe (struct addr:0xfffffa80021b6880)
[464] lsm.exe (struct addr:0xfffffa80021b90f0)
[556] svchost.exe (struct addr:0xfffffa800221a510)
[616] svchost.exe (struct addr:0xfffffa8001876060)
[716] svchost.exe (struct addr:0xfffffa8002268b30)
[764] svchost.exe (struct addr:0xfffffa8002291b30)
[788] svchost.exe (struct addr:0xfffffa8002298b30)
[948] svchost.exe (struct addr:0xfffffa80022fe9e0)
[260] svchost.exe (struct addr:0xfffffa8002331440)
[1044] spoolsv.exe (struct addr:0xfffffa80023a5b30)
[1084] dwm.exe (struct addr:0xfffffa80023ccb30)
[1092] taskhost.exe (struct addr:0xfffffa80023e0340)
[1108] explorer.exe (struct addr:0xfffffa80023e3b30)
[1168] svchost.exe (struct addr:0xfffffa800240f6e0)
[1332] svchost.exe (struct addr:0xfffffa800248e060)
[1380] sppsvc.exe (struct addr:0xfffffa80024bb060)
[1548] wlms.exe (struct addr:0xfffffa8002582630)
[1804] svchost.exe (struct addr:0xfffffa800238d3a0)
[1244] svchost.exe (struct addr:0xfffffa80011bfb30)
~~~

cc @smaresca ;)

noxdafox commented on tools/pyvmi/libvmi_cdef.h:
```diff
@@ -0,0 +1,339 @@
+} page_size_t;
+
+// page_info_t
+typedef struct page_info {
```
In case you don't need direct access to the struct fields, you can replace them with a literal "...". The compiler will fill the rest of the information for you. This makes the code more portable as, if the struct gets changed, you just need to compile it against the new header.

See: http://cffi.readthedocs.io/en/latest/overview.html#abi-versus-api

Wenzel replied:
Done, thanks

Wenzel commented:
So i have most of the functions implemented.

Othewise, a few details that needs to be solved:

- Do we want to have the full libvmi API there ?
Some functions like `vmi_print_hex` are not really useful since Python can already deal with this.

- some functions are missing:
 `convert_str_encoding`
 `vmi_free_unicode_str`
 `vmi_get_access_mode`

- The only way to initialize the library is by using a config file and passing the domain name to the constructor for now.

- functions like `vmi_read` requires an AccessContext *, we need to provide a wrapper on top of it, because `libvmi.py` should be the only one to `import ffi`.

- same thing for functions such as `vmi_get_page_mode`, which will return a CFFI `page_mode_t`, maybe we need  a wrapper.

if one of you could try it, and give me some feedback, it would be great.
Thanks.

Wenzel commented:
Actually, i just noticed that @bdpayne had written the same wrapper in #161 ...

I'll take some ideas there, and we can work on this PR i guess ?

Wenzel commented:
Update:
The last 2 commits added a support for:
- `VMI_CONFIG_STRING`
- `VMI_CONFIG_GHASHTABLE`
Now you need to build another cffi interface for `glib` with `python3 buld_glib.py`.

Examples:

`VMI_CONFIG_GLOBAL_FILE_ENTRY`
~~~Python
from libvmi import Libvmi

with Libvmi("Windows_7") as mi:
    os = vmi.get_ostype()
    ...
~~~


`VMI_CONFIG_STRING`
~~~Python
from libvmi import Libvmi, VMIConfig

config_str = '{ostype = "Windows";win_pdbase=0x28;win_pid=0x180;win_tasks=0x188;win_pname=0x2e0;}'

with Libvmi("Windows_7", mode=VMIConfig.STRING, config=config_str) as vmi:
    os = vmi.get_ostype()
    ...
~~~

`VMI_CONFIG_GHASHTABLE`
~~~Python
from libvmi import Libvmi, VMIConfig

hash = {
    "ostype": "Windows",
    "win_pdbase": 0x28,
    "win_tasks": 0x188,
    "win_pid": 0x180,
    "win_pname": 0x2e0,
}

with Libvmi("Windows_7", mode=VMIConfig.DICT, config=hash) as vmi:
    os = vmi.get_ostype()
    ...
~~~

cc @tklengyel what do you think of this interface ?
I'm happy to continue working on this python wrapper if you give some directions on what we should do next :)

Wenzel commented:
Hi @smaresca , i have been told that you could give some interesting feedback on this.
I you have a bit of time, i will really appreciate !

thanks.

noxdafox commented on tools/pyvmi/module-list.py:
```diff
@@ -0,0 +1,87 @@
+@contextmanager
+def pause(vmi):
+    vmi.pause_vm()
+    yield
```
On error here, the VM will be left paused. Is it intended?

If not the common `contextmanager` pattern is:

```python
@contextmanager
def pause(vmi):
    vmi.pause_vm()
    try:
        yield
    finally:
        vmi.resume_vm()
```

Wenzel replied:
I thought the contextmanager was already dealing with this try catch.
thanks, i will add it.

noxdafox commented on tools/pyvmi/module-list.py:
```diff
@@ -0,0 +1,87 @@
+from libvmi import Libvmi, VMIOS, PageMode
+
+
+def init_logger():
```
Both `init_logger` and `pause` are duplicated in 2 modules, maybe better to put them in a common module (common.py, utils.py)?

noxdafox commented on tools/pyvmi/libvmi.py:
```diff
@@ -0,0 +1,665 @@
+                # keep a reference to avoid GC
+                ghash_ref[key] = value
+
+            # replace config
```
TODO left here.

Wenzel replied:
fixed.

noxdafox commented on tools/pyvmi/libvmi.py:
```diff
@@ -0,0 +1,665 @@
+        status = lib.vmi_read(self.vmi, ctx, count, buffer, bytes_read)
+        check(status)
+        # transform into Python bytes
+        buffer = ffi.buffer(buffer, bytes_read[0])[:]
```
You can use the [`ffi.string`](http://cffi.readthedocs.io/en/latest/ref.html#ffi-string-ffi-unpack) interface here which is more neat. In this way you don't have to slice-copy the buffer.

Wenzel replied:
replaced, thanks.

noxdafox commented on tools/pyvmi/libvmi.py:
```diff
@@ -0,0 +1,665 @@
+        status = lib.vmi_read_ksym(self.vmi, symbol.encode(), count, buffer, bytes_read)
+        check(status)
+        # transform into Python bytes
+        buffer = ffi.buffer(buffer, bytes_read[0])[:]
```
Same as above and in few other places.

noxdafox commented on tools/pyvmi/libvmi.py:
```diff
@@ -0,0 +1,665 @@
+        self.free_unicode_str(value)
+        return buffer.decode(encoding)
+
+    # missing convert_str_encoding
```
TODO here?

Wenzel replied:
yes

noxdafox commented on tools/pyvmi/libvmi.py:
```diff
@@ -0,0 +1,665 @@
+            raise LibvmiError('VMI_FAILURE')
+        return ffi.string(value).decode()
+
+    # write
```
Redundant comment?

Wenzel replied:
it was just to separate all the read_* APIs from the write_* with a comment that should be view as a "section separator"

noxdafox commented on tools/pyvmi/libvmi_build.py:
```diff
@@ -0,0 +1,26 @@
+ffi.cdef(cdef_content)
+
+if __name__ == "__main__":
+    ffi.compile()
```
You might want to add the `verbose=True` flag here, it helps spotting issues at compile time.

noxdafox commented on tools/pyvmi/glib_build.py:
```diff
@@ -0,0 +1,45 @@
+ffi.cdef(cdef)
+
+if __name__ == "__main__":
+    ffi.compile()
```
You might want to add the `verbose=True` flag here, it helps spotting issues at compile time.

Wenzel replied:
maybe adding a command line flag `--debug` ?
Or having a verbose compilation is not a big deal ?

Wenzel replied:
Actually this will be called by the `setup.py`, so it's better to have a verbose output always enabled.

noxdafox commented on tools/pyvmi/setup.py:
```diff
@@ -1,21 +1,15 @@
+    version='2.0',
+    description='Python interface to LibVMI',
+    setup_requires=["cffi>=1.0.0", "pkgconfig"],
+    install_requires=["cffi>=1.0.0"],
```
Missing `docopt` dependency.

Wenzel replied:
`docopt` is only needed for the examples (`process-list` and `module-list`), but not for the `libvmi.py` module itself.

smaresca replied:
yes, i'd rather that it not be expressed as a dependency for the python layer. Regarding the examples, I think that it would be better to use more standard imports that are more likely to be available (already) on a user's system -- getopt itself may not be as featureful, but it is guaranteed to be present

noxdafox commented on tools/pyvmi/setup.py:
```diff
@@ -1,21 +1,15 @@
-      description = 'Python interface to LibVMI',
-      ext_modules = [pyvmimod])
+
+setup(
```
Your package won't include the external files meaning it won't compile on another host. You need to add them via the `data_files` keyarg.

You also probably need to include the `ext_modules` or `ext_package` keywords. Check the [`distutils`](https://docs.python.org/3.6/distutils/setupscript.html#describing-extension-modules) for more information.

I'd recommend you to try to run
```
$ python3 setup.py sdist
``` 
Then grab the resulting `tar.gz` and try to install on a separate host to test it builds correctly.

You can check a complete example [here](https://github.com/noxdafox/clipspy/blob/master/setup.py#L66).

Wenzel replied:
good catch !
i had to create a `libvmi/` package, and use `package_data` to include the `libvmi_cdef.h`

smaresca commented on tools/pyvmi/libvmi/libvmi_build.py:
```diff
@@ -0,0 +1,26 @@
+
+script_dir = os.path.dirname(os.path.realpath(__file__))
+# we read our C definitions from an external file
+# easier to maintain + C syntax highlighting
```
I believe it would be more maintainable to use libvmi/libvmi.h directly rather than making a shadow version of it in libvmi_cdef

Wenzel replied:
I have been advised the contrary: just pick the APIs that i'm really using in my cffi and copy paste their prototype.
Also, CFFI is not able to fully parse C headers as of today, so a slightest "exotic" C declaration could break `libvmi_build.py`.

But let me try.

Wenzel replied:
ping @noxdafox , can i have your opinion on CFFI ?

Wenzel replied:
Yep, it fails.

Wenzel commented:
A few things i have in mind:
- what do we need to upload a package on Pypi ?
- what about moving the code into it's own repository, like `python-libvmi`, instead of having it inside `tools/pyvmi` ?
we could have a independant travis test suite, and it's more maintanable anyway.

alex-bender commented:
@smaresca @noxdafox Hey guys! Is it possible to see this PR merged soon? It toot some time for me to realized that current py bindings are broken and pretty old. So it would be great to see this shiny new implementation.

smaresca commented:
@alex-bender missed your comment -- yes definitely.

@Wenzel there are a few conflicts that need fixing, but after that, one last review and then we can merge at long last

Wenzel commented:
I updated `pyvmiaddressspace.py` with the new libvmi interface, and changed it to use Rekall instead of Volatility.
I renamed it `vmiaddressspace.py` to have a more generic name.

tklengyel commented:
> I updated pyvmiaddressspace.py with the new libvmi interface, and changed it to use Rekall instead of Volatility.

Could we have an interface that works for both? Or one for each?

Wenzel commented:
That's actually a good question.
I will ask Michael Cohen, he helped get this implementation working.

A few things i need to check:
- is volatility python3 compatible ?
- is this cffi libvmi wrapper python 2 compatible ? (I hardcoded `#! /usr/bin/env python3` in `setup.py`, `libvmi_build.py` and `glib_build.py`.

Wenzel commented:
So, after some investigation:
- volatility is python2 only
- this python wrapper is now python2 compatible

i moved the `vmiaddressspace.py` to a `rekall/` directory.
Same thing with `volatility/` but the plugin is not working yet:
~~~
$ python2 vol.py -l vmi://nitro_win7x64 pslist
...
VMIAddressSpace: Must be first Address Space
...
~~~

Not sure what's happening ... ?

TODO:
- fix volatility address space
- check what to do with `pyvmifs`
- add a way to partially init libvmi (right now , `vmi_init_complete` is called, so you need a libvmi config file to run `rekall`... :/)
- update README

Wenzel commented:
I added support for a `partial` initialization in the constructor, by calling `vmi_init`.
Therefore it doesn't require any configuration, since we just want to read the physical memory in the case of `Volatility/Rekall`.

I also sent a mail on the Volatility mailing list for help regarding the address space.
Right now i have this output when i try to initialize it:
~~~
Volatility Foundation Volatility Framework 2.6
xc: error: Could not obtain handle on privileged command interface (2 = No such file or directory): Internal error
VMI_ERROR: Failed to open libxc interface.
LibVMI Version 0.11.0
LibVMI Driver Mode 1
--completed driver init.
--got id from name (nitro_win7x64 --> 7)
**set image_type = nitro_win7x64
--libvirt version 1003001
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "pmemaccess", "arguments": {"path": "/tmp/vmiSJkA5Q"}}'
--kvm: using custom patch for fast memory access
**set allocated_ram_size = 59700000, max_physical_address = 0x59700000
No suitable address space mapping found
Tried to open image as:
 MachOAddressSpace: mac: need base
 LimeAddressSpace: lime: need base
 WindowsHiberFileSpace32: No base Address Space
 WindowsCrashDumpSpace64BitMap: No base Address Space
 WindowsCrashDumpSpace64: No base Address Space
 HPAKAddressSpace: No base Address Space
 VMWareMetaAddressSpace: No base Address Space
 VirtualBoxCoreDumpElf64: No base Address Space
 QemuCoreDumpElf: No base Address Space
 VMWareAddressSpace: No base Address Space
 WindowsCrashDumpSpace32: No base Address Space
 SkipDuplicatesAMD64PagedMemory: No base Address Space
 WindowsAMD64PagedMemory: No base Address Space
 LinuxAMD64PagedMemory: No base Address Space
 AMD64PagedMemory: No base Address Space
 IA32PagedMemoryPae: No base Address Space
 IA32PagedMemory: No base Address Space
 OSXPmemELF: No base Address Space
 MachOAddressSpace: MachO Header signature invalid
 LimeAddressSpace: Invalid Lime header signature
 WindowsHiberFileSpace32: No xpress signature found
 WindowsCrashDumpSpace64BitMap: Header signature invalid
 WindowsCrashDumpSpace64: Header signature invalid
 HPAKAddressSpace: Location is not of file scheme
 VMWareMetaAddressSpace: Location is not of file scheme
 VirtualBoxCoreDumpElf64: ELF Header signature invalid
 QemuCoreDumpElf: ELF Header signature invalid
 VMWareAddressSpace: Invalid VMware signature: -
 WindowsCrashDumpSpace32: Header signature invalid
 SkipDuplicatesAMD64PagedMemory: Incompatible profile WinXPSP2x86 selected
 WindowsAMD64PagedMemory: Incompatible profile WinXPSP2x86 selected
 LinuxAMD64PagedMemory: Incompatible profile WinXPSP2x86 selected
 AMD64PagedMemory: Incompatible profile WinXPSP2x86 selected
 IA32PagedMemoryPae: No valid DTB found
 IA32PagedMemory: No valid DTB found
 OSXPmemELF: ELF Header signature invalid
 VMIAddressSpace: Must be first Address Space
 FileAddressSpace: Must be first Address Space
 ArmAddressSpace: No valid DTB found
~~~

Maybe @bdpayne can help ?

TODO:
- [x] Implement partial initialization for `Rekall`/`Volatility`
- [ ] Fix `Volatility` address space
- [ ] Check what do to with `pyvmifs`
- [ ] Update `README`

Thanks.

smaresca commented:
@Wenzel does the address space commit fix the issue you had described?

Wenzel commented:
@smaresca no, i just improved the Rekall one.

The Volatility one is still buggy for some reason.
If you know a Volatility developer, i'd like to have a chat with them, because i sent a mail on the mailing-list and got no replies.

Wenzel commented:
Btw, maybe the fix is already here: https://github.com/libvmi/libvmi/pull/520
??

Wenzel commented:
Rebased on Libvmi `master`

Wenzel commented:
I opened an issue on Volatility: https://github.com/volatilityfoundation/volatility/issues/498

Wenzel commented:
I just removed the Rekall vmiaddressspace, since it has been merged in the upsteam Rekall repo !! :100: 

About `pyvmifs`, i was reading the `README.pyvmifs`:
~~~
The guest VM's memory will show up as a regular file named "mem" inside
the mountpoint. When you're done, you can just unmount it.
~~~ 

Do we need to maintain this tool now that we have an integration into `Rekall` and (soon) `Volatility` ?

Wenzel commented:
Good news everyone.
I fixed the `Volatility` address space:
~~~
$ ./vol.py -l vmi://nitro_win7x64 --profile=Win7SP0x64 pslist
Volatility Foundation Volatility Framework 2.6
xc: error: Could not obtain handle on privileged command interface (2 = No such file or directory): Internal error
VMI_ERROR: Failed to open libxc interface.
LibVMI Version 0.11.0
LibVMI Driver Mode 1
--completed driver init.
--got id from name (nitro_win7x64 --> 4)
**set image_type = nitro_win7x64
--libvirt version 1003001
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "pmemaccess", "arguments": {"path": "/tmp/vmiOajj28"}}'
--kvm: using custom patch for fast memory access
**set allocated_ram_size = 59700000, max_physical_address = 0x59700000
--qmp: virsh -c qemu:///system qemu-monitor-command nitro_win7x64 '{"execute": "human-monitor-command", "arguments": {"command-line": "info registers"}}'
Offset(V)          Name                    PID   PPID   Thds     Hnds   Sess  Wow64 Start                          Exit                          
------------------ -------------------- ------ ------ ------ -------- ------ ------ ------------------------------ ------------------------------
0xfffffa8001108890 System                    4      0     76      455 ------      0 2018-03-12 22:37:30 UTC+0000                                 
0xfffffa8002210710 smss.exe                248      4      2       29 ------      0 2018-03-12 22:37:30 UTC+0000                                 
0xfffffa8002159060 csrss.exe               324    308      8      295      0      0 2018-03-12 22:37:30 UTC+0000                                 
0xfffffa8007b0c060 wininit.exe             372    308      3       74      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa8004dfd4c0 csrss.exe               380    364      7      141      1      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa80029d7870 winlogon.exe            420    364      3      108      1      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa8002956b30 services.exe            464    372      5      187      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa8002a133e0 lsass.exe               472    372      5      511      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa8002a0a7c0 lsm.exe                 480    372      9      135      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa80028b0b30 svchost.exe             588    464      9      336      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa80028b59e0 svchost.exe             648    464      7      214      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa80028ba9e0 svchost.exe             700    464     17      425      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa8002aa9b30 svchost.exe             776    464     17      409      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa80028c1640 svchost.exe             860    464     25      823      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa8002b4bb30 svchost.exe            1012    464     12      316      0      0 2018-03-12 22:37:31 UTC+0000                                 
0xfffffa8002b7b060 svchost.exe             368    464     16      395      0      0 2018-03-12 22:37:32 UTC+0000                                 
0xfffffa8002c954a0 spoolsv.exe            1036    464     12      264      0      0 2018-03-12 22:37:32 UTC+0000                                 
0xfffffa8002ccd740 svchost.exe            1076    464     17      320      0      0 2018-03-12 22:37:32 UTC+0000                                 
0xfffffa8002d6b4a0 svchost.exe            1208    464      8      152      0      0 2018-03-12 22:37:32 UTC+0000                                 
0xfffffa8002d771c0 taskhost.exe           1256    464      7      143      1      0 2018-03-12 22:37:32 UTC+0000                                 
0xfffffa8002de0a60 sppsvc.exe             1436    464      4      146      0      0 2018-03-12 22:37:32 UTC+0000                                 
0xfffffa8002e28060 wlms.exe               1512    464      4       41      0      0 2018-03-12 22:37:32 UTC+0000                                 
0xfffffa8002e99b30 svchost.exe            1752    464      5       90      0      0 2018-03-12 22:37:33 UTC+0000                                 
0xfffffa8002a30b30 dwm.exe                1832    776      3       68      1      0 2018-03-12 22:39:32 UTC+0000                                 
0xfffffa8002d8d060 explorer.exe            804   1772     19      608      1      0 2018-03-12 22:39:32 UTC+0000                                 
0xfffffa80010f30f0 svchost.exe            1636    464     12      337      0      0 2018-03-12 22:39:33 UTC+0000                                 
0xfffffa8001626530 svchost.exe            1468    464      6       66      0      0 2018-03-12 23:02:04 UTC+0000                                 
0xfffffa8001624b30 VSSVC.exe              1356    464      6      114      0      0 2018-03-12 23:07:33 UTC+0000  
~~~

There was 2 problems:
- Volatility was expecting a `VirtualAddressSpace` at the end of the selection process, and for that it has to use a `paged` address space, which requires  a `dtb` attribute.
- It requires also a `zread` method.

So the behavior is different from `Rekall`

TODO:
- [ ] Look at `Pyvmifs`
- [ ] Update `README` with the new interface
- [ ] Library name ? do we keep `PyVMI` ? `python-libvmi` ? or just `libvmi` ?

Personally, i would prefer to search for `libvmi`, it's generic and it's what you would be looking for in the first place.

tklengyel commented:
@Wenzel That's awesome! :) I would also give my +1 to just name it `libvmi`.

Wenzel commented:
So I updated the README, adding some examples to get started, and how to use the Volatility/Rekall integration.

About `pyvmifs` i can't even get the FUSE Python bindings installed...

As this PR is already quite big, I would vote for merging this one and opening another one to fix `pyvmifs` later.

tklengyel commented:
I think we can even remove `pyvmifs`, I don't see much value in it with the availability of `vmifs`.

Wenzel commented:
Ok i made some changes, removed `pyvmifs` and changed the name to `libvmi`.

I also implemented the `AccessContext` wrapper for CFFI.

Tell me what you think, and what's missing for the merge !

Wenzel commented:
the `astyle` travis check fails:
~~~
0.37s$ sudo add-apt-repository "deb mirror://mirrors.ubuntu.com/mirrors.txt xenial main restricted universe multiverse" -y
before_install.2
0.49s$ sudo apt-key adv --recv-keys --keyserver keyserver.ubuntu.com 3B4FE6ACC0B21F32
Executing: /tmp/tmp.U4vtOTMwIJ/gpg.1.sh --recv-keys
--keyserver
keyserver.ubuntu.com
3B4FE6ACC0B21F32
gpg: requesting key C0B21F32 from hkp server keyserver.ubuntu.com
gpg: key C0B21F32: "Ubuntu Archive Automatic Signing Key (2012) <ftpmaster@ubuntu.com>" 4 new signatures
gpg: Total number processed: 1
gpg:         new signatures: 4
3.03s$ sudo apt-get update -qq
W: http://ppa.launchpad.net/couchdb/stable/ubuntu/dists/trusty/Release.gpg: Signature by key 15866BAFD9BCC4F3C1E0DFC7D69548E1C17EAB57 uses weak digest algorithm (SHA1)
E: Failed to fetch mirror://mirrors.ubuntu.com/mirrors.txt/dists/xenial/InRelease  Clearsigned file isn't valid, got 'NOSPLIT' (does the network require authentication?)
E: Some index files failed to download. They have been ignored, or old ones used instead.
The command "sudo apt-get update -qq" failed and exited with 100 during .
~~~

tklengyel commented:
I saw that too, it can be ignored, seems to be an issue with the Travis system itself.

Wenzel commented:
Also, what do you think about moving the code to a separate repository under `libvmi` ?
Like `libvmi/python-libvmi`

As the code base will grow with examples, a wrapper for the event API and more, maybe it will be better to put i in it's own repository (with it's own Wiki for example)

Another thing will be managing the package upload to `pypi` a bit easier.

And tell me if i need to modify anything, i'm just waiting for the merge :)

tklengyel commented:
We can certainly move it to its own repo, I guess after this PR is closed. I would like to have @smaresca  give it a scan and merge it if he sees no outstanding items.

Wenzel commented:
Should we rename `libvmi/tools/pyvmi` to `libvmi/tools/python` to not confuse people ?

tklengyel commented:
Sure, I have no objections against renaming it, although if it will be moved into its on repository I don't see the point of changing it now.

Wenzel commented:
@tklengyel to not confuse users when this PR will be merged and the library has been renamed from `pyvmi` to `libvmi`.
After that, yes it doesn't matter

smaresca commented:
No need to rename it IMO, especially if we'll be extracting to a new repo (which I think is a rather good idea).  I'll re-review this in about an hour, and then we'll wrap up

smaresca commented:
Overall, the way this has progressed feels fairly pythonic to me. I like the use of context manager semantics, etc.  This will be a well used module in time, I've no doubt

Wenzel commented:
@smaresca how is the review going ?
Do i need to fix anything ?

smaresca commented:
@Wenzel no changes needed, I think it's good to go. Just one request: please squash everything. It's a giant pile of commits at the moment, and we'd prefer to keep the history concise.

Wenzel commented:
@smaresca done !

smaresca commented:
@Wenzel :) 

i'll set up the python-libvmi stuff we discussed

Wenzel commented:
A note: We need to rename `README` to `README.md` otherwise the Markdown is not interpreted
